std-bits/table/src/main/java/org/enso/table/util/ImmutableBitSet.java

```diff
@@ -0,0 +1,83 @@
+package org.enso.table.util;
+
+import java.util.BitSet;
+
+/**
+ * A wrapper around BitSet that implements boolean operations conveniently. Unlike BitSet,
+ * ImmutableBitSet takes a size parameter, which allows .not to be implemented.
+ */
+public class ImmutableBitSet {
+  private BitSet bitSet;
+  private int size;
+
+  public ImmutableBitSet(BitSet bitSet, int size) {
+    this.bitSet = bitSet;
+    this.size = size;
+  }
+
+  public BitSet toBitSet() {
+    return bitSet;
+  }
+
+  public ImmutableBitSet and(ImmutableBitSet other) {
+    assert size == other.size;
+    BitSet result = (BitSet) bitSet.clone();
+    result.and(other.bitSet);
+    return new ImmutableBitSet(result, size);
+  }
+
+  public ImmutableBitSet or(ImmutableBitSet other) {
+    assert size == other.size;
+    BitSet result = (BitSet) bitSet.clone();
+    result.or(other.bitSet);
+    return new ImmutableBitSet(result, size);
+  }
+
+  public ImmutableBitSet andNot(ImmutableBitSet other) {
+    assert size == other.size;
+    BitSet result = (BitSet) bitSet.clone();
+    result.andNot(other.bitSet);
+    return new ImmutableBitSet(result, size);
+  }
+
+  public ImmutableBitSet not() {
+    BitSet result = (BitSet) bitSet.clone();
+    result.flip(0, size);
+    return new ImmutableBitSet(result, size);
+  }
+
+  public ImmutableBitSet notAnd(ImmutableBitSet other) {
+    assert size == other.size;
+    BitSet result = (BitSet) bitSet.clone();
+    result.flip(0, size);
+    result.and(other.bitSet);
+    return new ImmutableBitSet(result, size);
+  }
+
+  public ImmutableBitSet notAndNot(ImmutableBitSet other) {
+    assert size == other.size;
+    BitSet result = (BitSet) bitSet.clone();
+    result.flip(0, size);
+    result.andNot(other.bitSet);
+    return new ImmutableBitSet(result, size);
+  }
+
+  public ImmutableBitSet orNot(ImmutableBitSet other) {
+    // Doing an extra operation to avoid doing an extra allocation.
+    // a || !b => !(!a && b)
+    assert size == other.size;
+    BitSet result = (BitSet) bitSet.clone();
+    result.flip(0, size);
+    result.and(other.bitSet);
+    result.flip(0, size);
+    return new ImmutableBitSet(result, size);
+  }
+
+  public static ImmutableBitSet allFalse(int size) {
+    return new ImmutableBitSet(new BitSet(), size);
```

Comment:
Is 

return new ImmutableBitSet(new BitSet(size), size);

going to make the allocation more efficient?

Reply:
Looking at the usages, I think to the contrary.

The thing is that `new BitSet()` creates an empty bitset, which will answer `false` to any `bitset.get(ix)`.

So it creates the `allFalse` bitset while using the minimal amount of memory. Whereas specifying the `size`, will pre-allocate the memory - which in this case just seems unnecessary.

So actually I think that keeping _not_ adding the `size` is best for allocation (but I'm only 85% sure about it).

Reply:
Yes, `BitSet` is really a set of 1-positions, not a set of boolean values, which is why I needed `ImmutableBitSet` to have a size field, so it would really be more like a set of bits. Specifically, you can't use `.not()` unless you know what the last bit position is. (The `not`-related methods are the only ones that use `size` for something other than checking sizes.